examples/inspection/plot_linear_model_coefficient_interpretation.py

```diff
@@ -0,0 +1,491 @@
+# The model learnt is far from being a good model making accurate predictions:
+# the R squared score is very low.
+# In the following section, we will interpret the coefficients of the model.
+# While we do so, we should keep in mind that any conclusion we way draw will
+# be about
+# the model that we build, rather than about the true (real-world) generative
+# process of the data.
+#
+# Interpreting coefficients
+# -------------------------
+#
+# First of all, we can plot the values of the coefficients of the regressor we
+# have fitted.
+
+feature_names = (model.named_steps['columntransformer']
+                      .named_transformers_['onehotencoder']
+                      .get_feature_names(input_features=categorical_columns))
+feature_names = np.concatenate(
+    [feature_names, binary_columns, numerical_columns])
+
+coefs = pd.DataFrame(
+    model.named_steps['transformedtargetregressor'].regressor_.coef_,
+    columns=['Coefficients'], index=feature_names
+)
+coefs.plot(kind='barh', figsize=(9, 7))
+plt.axvline(x=0, color='.5')
+plt.subplots_adjust(left=.3)
+
+###############################################################################
+# Soon we realize that we cannot compare different coefficients since the
+# features have different natural scales and hence value ranges
+# because of their different unit of measure.
+# For instance, the AGE coefficient is expressed in $/hours/living years
+# while the EDUCATION one is expressed in $/hours/years of education.
+# Looking at the coefficient plot to extrapolate feature importance could be
+# misleading as some of them vary on a small scale (as UNION or SEX that are
+# either 0 or 1), while feature like AGE varies a lot more, several decades.
+# This is evident if we compare feature standard deviations.
+
+X_train_preprocessed = pd.DataFrame(
+    model.named_steps['columntransformer'].transform(X_train),
+    columns=feature_names
+)
+X_train_preprocessed.std(axis=0).plot(kind='barh', figsize=(9, 7))
+plt.title('Features std. dev.')
+plt.subplots_adjust(left=.3)
+
+###############################################################################
+# For the reasons explained above, multiplying the coefficients by the
+# standard deviation of the related feature would improve our understanding on
+# feature importance on the model.
+# In that way, we emphasize that the
+# greater the variance of a feature, the larger the weight of the corresponding
+# coefficient on the output, all else being equal.
+
+coefs = pd.DataFrame(
+    model.named_steps['transformedtargetregressor'].regressor_.coef_ *
+    X_train_preprocessed.std(axis=0),
+    columns=['Coefficient importance'], index=feature_names
+)
+coefs.plot(kind='barh', figsize=(9, 7))
+plt.axvline(x=0, color='.5')
+plt.subplots_adjust(left=.3)
+
+###############################################################################
+# The plot above tells us about dependencies between a specific feature and
```

Comment:
This paragraph should be about the first plot, not this one. The conclusion for this last plot should be different in order to better motivate the multiplication by the std dev, which despite all your efforts is still obscure to me :/

CC @GaelVaroquaux  

----

@cmarmo I know I've made this comment before and I know you acknowledged it, but since github has removed them, I'm just making sure they are here again so this doesn't get merged too fast. This is not me nagging you

Reply:
@NicolasHug this time the comment comes after a text modification.
It is more clear to me, now, where we disagree.
You are right the first plot already shows the behavior of the coefficients: but what it is worth to say here is that we do not want to come to a conclusion before having taken into account the possible pitfalls in the interpretation of linear coefficients. Once the variance of the feature has been taken into account then we can correctly interpret the result.
That's why I think that this comment is at his place here even if some conclusions are already clear in the previous plot.
The very first version, that I agreed to modify, was pointing too much on the feature trend than on the counter-linearity.
That said, if you think this not acceptable I will just change it in order to move forward.

Reply:
I have 2 problems with the current state of things:


1. I still don't understand why we need to multiply by the std dev

2. The following conclusions:
   > An increase of the AGE will induce a decrease of the WAGE when all other features remain constant. On the contrary, an increase of the EXPERIENCE will induce an increase of the WAGE when all other features remain constant

   can be reached by just looking at the coefficients values. We don't need to scale the coefficients by anything to deduce this. Saying this at the end, i.e. after having scaled the coeffs, wrongly suggests that we need to scale the coeff before we can draw such conclusion. That is not true and will confuse readers.

This is why I suggest that the conclusions for the last plot should be different. Hopefully it will also help me solve 1.

We need something like "Now that the coefficients have been scaled, we can draw conclusion X that we could not have made if we didn't scale the coeffs."  Or "we might have been tempted to draw conclusion Y before scaling the coeffs, but now that the coeffs are scaled, we realize that instead blah blah...."

(I'm not suggesting to write this verbatim, this is just the kind of logic that should be there.)

Reply:
Ok, I will reformulate the entire paragraph, let's see if I can do better.

Reply:
@NicolasHug I'm going to resolve this. Feel free to comment on the new version of the text. Thanks.

Reply:
I'm sorry @cmarmo  but this comment is still not addressed.

> An increase of the AGE will induce a decrease
 of the WAGE when all other features remain constant. On the contrary, an
 increase of the EXPERIENCE will induce an increase of the WAGE when all
 other features remain constant.

This should be said about the first plot. We don't need to have scaled coefficients for this to hold.

Reply:
I'm sorry @NicolasHug but I have explained why I think this sentence could stay there.
So I will stop to argue now. :) But I will not change this.
Feel free to close the PR and open a new one with your text.

Reply:
@cmarmo, the current text is claiming that we can do X only after we have done Y. That is simply not true: we don't need to scale the coeffs by the stddev to know whether they positively or negatively affect the target. This isn't an opinion of mine, it's just a fact. I might have done a poor job at conveying this point before, but I'm happy to further detail / discuss this with you if you'd like.

OTOH, having clear cause-to-effect relationships is an absolute necessity for such entry-level examples, where the target audience is beginners for which ML concepts aren't necessarily sharp yet. It is our responsibility to not be misleading. The current statement *is* misleading, unfortunately. 

I'm sorry that you choose not to address this comment and refuse further discussion. I hold strong opinions about the quality of our documentation, and I'm very attached to it. I hope we'll find common grounds.

Reply:
I think the point is to treat one problem after the next.

- First the problem of magnitudes of the coefficients that are not comparable unless the coefficients have the same unit;
- Second the problem of the sign of the coefficients that can be misleading if interpreted without taking correlated variables into account.

If we discuss all the pitfalls at the same time in the first plot the discussion will be too hard to follow. Maybe the title of the subsections could be improved to better define the scope of each part of the example.

- Pitfall #1: misleading magnitude of the coefficients on unscaled data

- Pitfall #2: unstable magnitude of the coefficients with correlated variables

- Pitfall #3: misleading sign of coefficients with correlated variables

- Pitfall #4: impact of regularization on the sign and magnitude of coefficients
 

Also we could put a Table of Content a the top of the example to let the reader that we will treat one problem after the next why letting the user know the list of potential issues ahead of time (sneak peek).

Reply:
I love @ogrisel 's suggestion.

For the record, here is the kind of change that I was advocating for. The only major change with the current version is that the interpretation about the coefficients absolute effect on the target is done right after the first plot, while the comparison of their relative value is left for the scaled plot.

Diff looks like a lot, but it's mostly moving sections around.

```diff
diff --git a/examples/inspection/plot_linear_model_coefficient_interpretation.py b/examples/inspection/plot_linear_model_coefficient_interpretation.py
index 1f24c70439..ed8f58c22c 100644
--- a/examples/inspection/plot_linear_model_coefficient_interpretation.py
+++ b/examples/inspection/plot_linear_model_coefficient_interpretation.py
@@ -223,22 +223,20 @@ plt.axvline(x=0, color='.5')
 plt.subplots_adjust(left=.3)
 
 ###############################################################################
-# In the plot above, the AGE coefficient is expressed in
+# The plot above tells us about dependencies between a specific feature and
+# the target when all other features remain constant, i.e., **conditional
+# dependencies**. An increase of the AGE by 1 will induce a decrease
+# of the WAGE by `coeff_age`, when all other features remain constant. On the
+# contrary, an increase of the EXPERIENCE by 1 will induce an increase of the
+# WAGE by `coeff_exp` when all other features remain constant.
+#
+# Note however that the features have vastly different scales:
+# the AGE coefficient is expressed in
 # :math:`$/hours/(living\ years)` while the EDUCATION one is expressed
 # in :math:`$/hours/(years\ of\ education)`.
 # On the other hand, categorical variables (as UNION or SEX) are adimensional
 # numbers taking the value either of 0 or 1. Their coefficients are expressed
-# in :math:`$/hours`. An increase of
-# :math:`1` in AGE is not comparable with an increase of :math:`1` in UNION.
-# We cannot compare different coefficients since the
-# features have different natural scales and hence value ranges
-# because of their different unit of measure.
-# Indeed, from this plot the most important factor in determining WAGE is the
-# variable UNION, even if it is plausible that variables like EXPERIENCE
-# should have more impact.
-# Looking at the coefficient plot to extrapolate feature importance could be
-# misleading as some of them vary on a small scale, while others, like AGE,
-# varies a lot more, several decades.
+# in :math:`$/hours`.
 # This is evident if we compare feature standard deviations.
 
 X_train_preprocessed = pd.DataFrame(
@@ -250,15 +248,25 @@ plt.title('Features std. dev.')
 plt.subplots_adjust(left=.3)
 
 ###############################################################################
-# Multiplying the coefficients by the standard deviation of the related
-# feature would reduce all the coefficients to the same unit of measure.
-# As we will see :ref:`after<scaling_num>` this is equivalent to normalize
-# numerical variables to their standard deviation,
-# as :math:`y = \sum{coeff_i * X_i} = \sum{(coeff_i * std_i) * (X_i / std_i)}`.
+# As a result, an increase of
+# :math:`1` in AGE is not directly comparable with an increase of :math:`1` in
+# UNION.
+# We cannot compare different coefficients since the
+# features have different natural scales and hence value ranges
+# because of their different unit of measure.
+# Indeed, from our first coefficient plot, the most important factor in
+# determining WAGE seems to be the variable UNION, even if it is plausible that
+# variables like EXPERIENCE should have more impact.
+# Looking at the coefficient plot to extrapolate feature importance could be
+# misleading as some of them vary on a small scale, while others, like AGE,
+# varies a lot more, several decades.
 #
-# In that way, we emphasize that the
-# greater the variance of a feature, the larger the weight of the corresponding
-# coefficient on the output, all else being equal.
+# To better compare coefficient, one option is to preprocess the features by
+# dividing them by their standard deviation. This reduces all the features to
+# the same a-dimensional unit of measure. Alternatively, as we fitted our model
+# without pre-scaling, we can multiply the coefficients by the std-dev for an
+# equivalent result:
+# :math:`y = \sum{coeff_i * X_i} = \sum{(coeff_i * std_i) * (X_i / std_i)}`.
 
 coefs = pd.DataFrame(
     model.named_steps['transformedtargetregressor'].regressor_.coef_ *
@@ -271,21 +279,14 @@ plt.axvline(x=0, color='.5')
 plt.subplots_adjust(left=.3)
 
 ###############################################################################
-# .. warning::
-#
-#   Why does the plot above suggest that an increase in age leads to a
-#   decrease in wage? Why this is different from the :ref:`initial pairplot
-#   <marginal_dependencies>`?
+# Now that the coefficients have been scaled, we can start comparing their
+# relative values. In this plot, a coefficient represents the expected
+# difference in the predicted value, when its corresponding variable differs by
+# one standard deviation, while all other features remain fixed at their average
+# values.
 #
-# Now that the coefficients have been scaled, we can start to interpret them.
-# The plot above tells us about dependencies between a specific feature and
-# the target when all other features remain constant, i.e., **conditional
-# dependencies**. An increase of the AGE will induce a decrease
-# of the WAGE when all other features remain constant. On the contrary, an
-# increase of the EXPERIENCE will induce an increase of the WAGE when all
-# other features remain constant.
-# Also, AGE, EXPERIENCE and EDUCATION are the three variables that most
-# influence the model.
+# It appears now that AGE, EXPERIENCE and EDUCATION are the three
+# variables that most influence the model.
 #
 # Checking the variability of the coefficients
 # --------------------------------------------
```